### ISSUE TITLE
Support CSV via Type Manifold

### ISSUE THREAD

**Issue by rsmckinney** [manifold-systems/manifold#113]
Support CSV via Type Manifold

Natalie is working on the CSV parser and type inference

**Comment by wxq1990** [comment]
expecting

**Comment by rsmckinney** [comment]
it’s coming, ETA end of year

**Comment by rsmckinney** [comment]
in progress...

**Comment by rsmckinney** [comment]
Pre-implementation notes in no particular order:

CSV manifold
- basic format https://tools.ietf.org/html/rfc4180#page-2
- see the apache csv library for info about formats (https://stackabuse.com/reading-and-writing-csvs-in-java-with-apache-commons-csv/)

NAMING
- given file resources/com/example/Foo.csv, the user accesses it using the following name:
-- com.example.Foo

- HEADER:
our parser infers the header row's existence via sampling

DIALECTS
- our parser infers the separator character (comma, semi, tab) via sampling, only one separator may be used in a given file
- our parser ignores empty lines (excel and other dialects do this too)
- our parser allows empty data and empty header column via ,,
- our parser's treatment of double quotes
-- if a field begins with a " it must end with a " i.e., a leading " implies quoted data, 
   otherwise " is a legal character in the data
-- (rfc4180) any character including line break is legal inside quoted data, where a " must be escaped as two ""
- our parser infers the treatment of leading and trailing spaces via sampling
-- if the separator is consistently followed by and/or preceded by the same number of spaces, leading/trailing spaces are ignored unless quoted,
   otherwise leading/trailing spaces are preserved in unquoted data

DATA TYPES
- our parser infers data types? (maybe)
-- possibilities: string, integer/big, decimal/big, boolean, date-time

OVERRIDE/CONFIG
- provide means to configure parser's behavior:
-- force header (force first line as header)
-- provided header (allow header to be provided separately)
-- force delimiter (specify delimiter)
-- force quotes (all quotes are literal)
-- force whitespace (all spaces significant)
-- data types provided

INTERCHANGEABLE
- XML will parse as `DataBindings` which makes it interchangeable with JSON, XML, & YAML
- Load JSON, XML, YAML from CSV 
- Use JSON Schema as CSV Schema
-- com/example/MyType.json defines a type for some CSV files
-- MyType foo = MyType.load().fromCsv(urlFooCsvFile)

**Comment by sporkisfaster** [comment]
Not sure if this matters or still isn't implemented but in tinkering with this manifold cannot seem to find .csv files when using fromSource()

**Comment by rsmckinney** [comment]
It matters. Can you provide steps to reproduce?

**Comment by sporkisfaster** [comment]
Sure. 

- Created new project in IntelliJ

- Created simple CSVTest.java class

- Dropped a .csv file in a resource folder (src/main/resources/abc/csv)

- Imported the appropriate class as well as class item (import abc.csv.TestScenario as well as TestScenarioItem)

- Attempt to instantiate a TestScenario object called test 
 ```java
    public static void manifoldTest() {
        TestScenario test = TestScenario.fromSource();
        for(TestScenarioItem tests: test) {
            System.out.println(tests.getLOAD());
        }
    }
```

- Upon run or compile, errors stating that java cannot find symbol TestScenarioItem appear
<img width="996" alt="Screen Shot 2019-11-19 at 8 58 56 AM" src="https://user-images.githubusercontent.com/3966216/69157661-de637700-0aaa-11ea-809c-1954a8175211.png">

Manifold-all version: 2019.1.28
Plugin Version: 2019.2.6
IJ Version: 2019.2.4

**Comment by sporkisfaster** [comment]
Also, and this may be wholly unrelated but worth mentioning: I keep getting prompted by IntelliJ that my libraries are out of date even though I definitely have latest plugin and library
<img width="579" alt="Screen Shot 2019-11-19 at 10 32 49 AM" src="https://user-images.githubusercontent.com/3966216/69165999-23da7100-0ab8-11ea-9a90-cffecef965a7.png">

**Comment by rsmckinney** [comment]
Whoops!  I changed the plugin's version number scheme recently, there are three separate plugin releases for 2019.1 ,2019.2, & 2019.3 (recent IJ API changes made it too difficult to manage one release for all versions).  Anyway, the release numbers no longer align with the logic to detect the latest version of manifold.... I'll fix this ASAP.  Thanks for reporting this, Wes!

**Comment by rsmckinney** [comment]
Are you certain your project is using Manifold 2019.1.28?  I can't reproduce this one.

**Comment by sporkisfaster** [comment]
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <profiles>
        <profile>
            <id>internal.tools-jar</id>
            <activation>
                <file>
                    <exists>${java.home}/../lib/tools.jar</exists>
                </file>
            </activation>
            <dependencies>
                <dependency>
                    <groupId>com.sun</groupId>
                    <artifactId>tools</artifactId>
                    <version>1.8.0</version>
                    <scope>system</scope>
                    <systemPath>${java.home}/../lib/tools.jar</systemPath>
                </dependency>
            </dependencies>
        </profile>
    </profiles>

    <groupId>csv</groupId>
    <artifactId>manifoldtests</artifactId>
    <version>0.1-SNAPSHOT</version>

    <properties>
        <!-- set latest manifold version here -->
        <manifold.version>2019.1.28</manifold.version>
        <maven.compiler.source>8</maven.compiler.source>
        <maven.compiler.target>8</maven.compiler.target>
    </properties>

    <dependencies>
        <dependency>
            <groupId>systems.manifold</groupId>
            <artifactId>manifold-all</artifactId>
            <version>${manifold.version}</version>
        </dependency>
    </dependencies>

    <!--Add the -Xplugin:Manifold argument for the javac compiler-->
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.0</version>
                <configuration>
                    <encoding>UTF-8</encoding>
                    <compilerArgs>
                        <arg>-Xplugin:Manifold</arg>
                    </compilerArgs>

                </configuration>
            </plugin>

        </plugins>
    </build>
</project>
```
Here's my pom
<img width="346" alt="Screen Shot 2019-11-19 at 2 03 10 PM" src="https://user-images.githubusercontent.com/3966216/69181782-5777c400-0ad5-11ea-900a-75b4f6d55267.png">

**Comment by rsmckinney** [comment]
Re ` TestScenarioItem`, it may relate to the data in the file.  Can you provide that somehow?

**Comment by rsmckinney** [comment]
RE
>Here's my pom
> Screen Shot 2019-11-19 at 2 03 10 PM

Maybe refresh the project via Maven tool window?

**Comment by sporkisfaster** [comment]
> Re ` TestScenarioItem`, it may relate to the data in the file. Can you provide that somehow?

Unfortunately not easily, but how do you mean so I can cross check in the file?

> RE
> 
> > Here's my pom
> > Screen Shot 2019-11-19 at 2 03 10 PM
> 
> Maybe refresh the project via Maven tool window?

Same thing happens.

**Comment by sporkisfaster** [comment]
> Re ` TestScenarioItem`, it may relate to the data in the file. Can you provide that somehow?

I think this is it. I tried the insurance csv in the csv test repo and it worked. What are the preconditions to using that item interface?

**Comment by rsmckinney** [comment]
>Unfortunately not easily, but how do you mean so I can cross check in the file?

I suspect a bug related to sampling the data or similar, so would be difficult to diagnose without the data. Perhaps provide a single record of the CSV file with alpha text changed to random chars?

**Comment by rsmckinney** [comment]
Also, if there is a header row, that would be important too.

**Comment by rsmckinney** [comment]
Re the warning message about the version, I think I see the problem  (on my end). I'll reproduce this and then fix ASAP.  Thanks for reporting this.

**Comment by sporkisfaster** [comment]
> > Unfortunately not easily, but how do you mean so I can cross check in the file?
> 
> I suspect a bug related to sampling the data or similar, so would be difficult to diagnose without the data. Perhaps provide a single record of the CSV file with alpha text changed to random chars?

Let me have a look. Thanks for the help!

**Comment by rsmckinney** [comment]
Thanks for you cooperation, much appreciated.

**Comment by rsmckinney** [comment]
Found and fixed issue re manifold version check.  Will be available in next plugin release, end of this week.  Thanks again for reporting.

I'll work on the CSV fix when you have time to provide some data.  Thanks!

**Comment by rsmckinney** [comment]
@sporkisfaster Wes, just fyi, I'm going to log a separate issue for the manifold version check so I can track that separately.

**Comment by rsmckinney** [comment]
@sporkisfaster 
>Not sure if this matters or still isn't implemented but in tinkering with this manifold cannot seem to find .csv files when using fromSource()

I should have asked this earlier. Were you able to load the .csv using a different method? e.g., fromFile()?  Also in IJ when you try to use type, does it reflect the header properly?

**Comment by rsmckinney** [comment]
@sporkisfaster Fyi, your original issue with fromSource() may be fixed now with latest manifold release [2019.1.29](https://github.com/manifold-systems/manifold/releases/tag/v2019.1.29).  Please verify when you have time.  Thanks!
